File: code/tests/common/vehicle.py
import numpy as np


def get_heading(theta):
    return np.arctan2(np.sin(theta), np.cos(theta))


class Vehicle:
    def __init__(self, x=0, y=0, velocity=0, heading=0, wheel_base=2):
        self.x = x
        self.y = y
        self.velocity = velocity
        self.heading = heading
        self.wheel_base = wheel_base

    def update(self, acc, dt, delta):
        self.x += self.velocity * np.cos(self.heading) * dt
        self.y += self.velocity * np.sin(self.heading) * dt
        ang_vel = self.velocity * np.tan(delta) / self.wheel_base
        self.heading = get_heading(self.heading + ang_vel * dt)
        self.velocity += acc * dt

File: game/src/models/vehicle.ts
export class Vehicle {
    public location : {x: number, y: number}
    public heading : number
    public velocity : number
    public wheel_base : number
    public mass : number

    constructor(x = 0, y = 0, v = 0) {
        this.location = {x, y}
        this.heading = 0
        this.velocity = v
        this.wheel_base = 2
        this.mass = 100
    }

    private getHeading(theta){
        return Math.atan2(Math.sin(theta), Math.cos(theta))
    }

    public update(acc, delta, dt){
        this.location.x += this.velocity * Math.cos(this.heading) * dt
        this.location.y += this.velocity * Math.sin(this.heading) * dt
        const ang_vel = this.velocity * Math.tan(delta) / this.wheel_base
        this.heading = this.getHeading(this.heading + ang_vel * dt)
        const acc_eff = acc - this.velocity**2 * 0.001
        this.velocity += acc_eff * dt
    }
}